mono/utils/mono-conc-hashtable.c

```diff
@@ -351,12 +351,14 @@ mono_conc_hashtable_insert (MonoConcurrentHashTable *hash_table, gpointer key, g
 				kvs [i].value = value;
 				/* The write to values must happen after the write to keys */
 				mono_memory_barrier ();
-				kvs [i].key = key;
+				
 				if (kvs [i].key == TOMBSTONE)
 					--hash_table->tombstone_count;
 				else
 					++hash_table->element_count;	
-					
+				
+				kvs [i].key = key;
```

Comment:
@hecklatcipsoft From your description I see that the issue you are encountering comes from not resetting the tombstone / element_counts after the rehash. Isn't that fix enough ? In what case is this commit needed ?

Reply:
HI @BrzVlad ,

Well it's just plain wrong to assign to the key before checking key == tombstone, because it never will be after assignment.

But yes both commits are fixes for scenarios where lookups of elements that are not contained in the hashmap can result in an endless loop. 

simplified example (assume hashmap size: 4):
insert #1  => insert #2  => insert #3 ( tombstone_count= 0, element_count =3)
remove #1  => remove #2 ( tombstone_count= 2, element_count =3 )
insert #1  => insert #2 ( rehash() will rehash to size 4, tombstone_count= 2, element_count =5 )
remove #1  => remove #2 ( tombstone_count= 4, element_count =5 )
insert #1  => insert #2 ( rehash() will rehash to size 4, tombstone_count= 4, element_count =7 )
insert #4 ( rehash() will rehash to size 4, tombstone_count= 4, element_count =8 )
lookup #5 ( endless loop )

The second fix will effectively fix some of the errors introduced by the key assignment bug (including the simple example above). But I'm not sure if a scenario exists where the erroneous tombstone count will introduce other problems. So to be on the save side I recommend merging the key assignment fix as well.

Best, Jochen

Reply:
Wow, sorry. I didn't even notice that it was checking for the previously assigned value. Can't believe anything worked